gpu_time/gpu_time_test.cpp

```diff
@@ -0,0 +1,434 @@
+
+    // Simulate a 20ms duration for the second command buffer (queries 2 and 3)
+    // 20ms = 20,000,000 ns
+    timestamps[4] = 2000000000;  // Start time for cmd 2
+    timestamps[5] = 1;           // Availability
+    timestamps[6] = 2020000000;  // End time for cmd 2
+    timestamps[7] = 1;           // Availability
+
+    // Simulate a 30ms duration for the third command buffer (queries 4 and 5)
+    // 30ms = 30,000,000 ns
+    timestamps[8] = 3000000000;   // Start time for cmd 3
+    timestamps[9] = 1;            // Availability
+    timestamps[10] = 3030000000;  // End time for cmd 3
+    timestamps[11] = 1;           // Availability
+
+    return VK_SUCCESS;
+}
+
+VKAPI_ATTR VkResult VKAPI_CALL MockQueueWaitIdle(VkQueue queue)
+{
+    // No-op for testing
+    return VK_SUCCESS;
+}
+
+VKAPI_ATTR VkResult VKAPI_CALL MockDeviceWaitIdle(VkDevice device)
+{
+    // No-op for testing
+    return VK_SUCCESS;
+}
+
+void CreateGPUTime(Dive::GPUTime& gpuTime, float timestamp_period)
+{
+    ASSERT_TRUE(gpuTime
+                .OnCreateDevice(MOCK_DEVICE,
+                                /*allocator=*/nullptr,
+                                timestamp_period,
+                                MockCreateQueryPool,
+                                MockResetQueryPool)
+                .success);
+}
+
+void DestroyGPUTime(Dive::GPUTime& gpuTime)
+{
+    VkQueue queue = MOCK_QUEUE;
+    gpuTime.OnGetDeviceQueue(&queue);
+    ASSERT_TRUE(
+    gpuTime.OnDestroyDevice(MOCK_DEVICE, MockQueueWaitIdle, MockDestroyQueryPool).success);
+}
+
+void ExpectStatsEq(const Dive::GPUTime::Stats& stats,
+                   double                      expected_average,
+                   double                      expected_median,
+                   double                      expected_min,
+                   double                      expected_max,
+                   double                      expected_stddev)
+{
+    EXPECT_DOUBLE_EQ(stats.average, expected_average);
+    EXPECT_DOUBLE_EQ(stats.median, expected_median);
+    EXPECT_DOUBLE_EQ(stats.min, expected_min);
+    EXPECT_DOUBLE_EQ(stats.max, expected_max);
+    EXPECT_DOUBLE_EQ(stats.stddev, expected_stddev);
+}
+
+// --- GPUTime Tests ---
+// Test that a newly created object returns default statistics.
+TEST(GPUTimeTest, InitialStateReturnsDefaultStats)
+{
+    Dive::GPUTime gpuTime;
+    CreateGPUTime(gpuTime, kMockTimestampPeriod);
+
+    auto stats = gpuTime.GetStats();
+    ExpectStatsEq(stats,
+                  /* average */ 0.0,
+                  /* median */ 0.0,
+                  /* min */ std::numeric_limits<double>::max(),
+                  /* max */ std::numeric_limits<double>::lowest(),
+                  /* stddev */ 0.0);
+
+    DestroyGPUTime(gpuTime);
+}
+
+// Test that a command buffer can be successfully allocated.
+TEST(GPUTimeTest, AllocateCommandBufferSucceeds)
+{
+    Dive::GPUTime gpuTime;
+    CreateGPUTime(gpuTime, kMockTimestampPeriod);
```

Comment:
Assertions in subroutines do not halt the test. Consider using `ASSERT_NO_FATAL_FAILURE` or some alternative: https://google.github.io/googletest/advanced.html#asserting-on-subroutines

```suggestion
    ASSERT_NO_FATAL_FAILURE(CreateGPUTime(gpuTime, kMockTimestampPeriod));
```

Reply:
good to know, thanks for mentioning this. fixed